File: src/main/java/red/tetracube/smarthub/controllers/bot/BotController.java
package red.tetracube.smarthub.controllers.bot;

import io.smallrye.mutiny.Uni;
import red.tetracube.smarthub.controllers.bot.payloads.BotDeviceFeatureItem;
import red.tetracube.smarthub.controllers.bot.payloads.DeviceFeatureAction;

import javax.inject.Inject;
import javax.ws.rs.*;

import java.util.List;

import static javax.ws.rs.core.MediaType.APPLICATION_JSON;

@Path("/bot")
public class BotController {

    @Inject
    BotService botService;

    @GET
    @Path("/features")
    @Produces(APPLICATION_JSON)
    public Uni<List<BotDeviceFeatureItem>> getDevices() {
        return botService.getDeviceFeatures();
    }

    @GET
    @Path("/devices/{deviceName}/features/{featureName}/commands")
    @Produces(APPLICATION_JSON)
    public Uni<DeviceFeatureAction> getDeviceFeatureActions(@PathParam("deviceName") String deviceName,
                                                            @PathParam("featureName") String featureName) {
        return botService.getDeviceFeatureActions(deviceName, featureName)
                .map(optionalResponse -> {
                    if (optionalResponse.isEmpty()) {
                        throw new NotFoundException("Cannot find actions for specified device or feature name");
                    }
                    return optionalResponse.get();
                });
    }
}

File: src/main/kotlin/iot/tetracubered/bot/BotController.kt
package iot.tetracubered.bot

import io.smallrye.mutiny.Uni
import io.smallrye.mutiny.coroutines.asUni
import iot.tetracubered.bot.payloads.DeviceFeature
import iot.tetracubered.bot.payloads.DeviceFeatureAction
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.async
import javax.ws.rs.*
import javax.ws.rs.core.MediaType.APPLICATION_JSON

@ExperimentalCoroutinesApi
@Path("/api/bot")
class BotController(
    private val botServices: BotServices
) {

    @GET
    @Path("/features")
    @Produces(APPLICATION_JSON)
    fun getDevicesList(): Uni<List<DeviceFeature>> {
        return GlobalScope.async {
            botServices.getDevices()
        }.asUni()
    }

    @GET
    @Path("/devices/{deviceName}/features/{featureName}/commands")
    @Produces(APPLICATION_JSON)
    fun getDeviceFeatureActions(
        @PathParam("deviceName") deviceName: String,
        @PathParam("featureName") featureName: String
    ): Uni<DeviceFeatureAction> {
        return GlobalScope.async {
            val response = botServices.getDeviceFeatureActions(deviceName, featureName)
                ?: throw NotFoundException("Cannot find actions for specified device or feature name")
            response
        }.asUni()
    }
}